torch/csrc/cuda/nccl.cpp

```diff
@@ -121,232 +129,330 @@ static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tenso
   }
 }
 
-static ncclDataType_t _get_data_type(at::TypeID type) {
-  switch(type) {
-  case at::TypeID::CUDAFloat   : return ncclFloat;
-  case at::TypeID::CUDAHalf    : return ncclHalf;
-  case at::TypeID::CUDADouble  : return ncclDouble;
-  case at::TypeID::CUDALong    : return ncclInt64;
-  case at::TypeID::CUDAInt     : return ncclInt;
-  case at::TypeID::CUDAChar    : return ncclChar;
-  case at::TypeID::CUDAByte    : return ncclChar;
+static ncclDataType_t _get_data_type(const Type& type) {
+  if (type.backend() != kCUDA) {
+    throw std::runtime_error("Unconvertible NCCL type");
+  }
+  switch (type.scalarType()) {
+  case at::kFloat   : return ncclFloat;
+  case at::kHalf    : return ncclHalf;
+  case at::kDouble  : return ncclDouble;
+  case at::kLong    : return ncclInt64;
+  case at::kInt     : return ncclInt;
+  case at::kChar    : return ncclChar;
+  case at::kByte    : return ncclChar;
   default: throw std::runtime_error("Unconvertible NCCL type");
   }
 }
 
+static void _start_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupStart());
+#endif
+}
+
+static void _end_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupEnd());
```

Comment:
Not sure what are the semantics of `ncclGroupEnd`, but if it should be called even if one of the collective calls fails, then we should make a scope guard with it.